File: mrgeo-mapalgebra/mrgeo-mapalgebra-image/src/main/java/org/mrgeo/mapalgebra/ChangeClassificationMapOp.java
/*
 * Copyright 2009-2014 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package org.mrgeo.mapalgebra;

import org.mrgeo.aggregators.Aggregator;
import org.mrgeo.aggregators.AggregatorRegistry;
import org.mrgeo.aggregators.MeanAggregator;
import org.mrgeo.buildpyramid.BuildPyramidSpark;
import org.mrgeo.data.DataProviderFactory;
import org.mrgeo.data.DataProviderFactory.AccessMode;
import org.mrgeo.data.image.MrsImageDataProvider;
import org.mrgeo.image.MrsImagePyramid;
import org.mrgeo.image.MrsImagePyramidMetadata;
import org.mrgeo.image.MrsImagePyramidMetadata.Classification;
import org.mrgeo.mapalgebra.parser.ParserAdapter;
import org.mrgeo.mapalgebra.parser.ParserNode;
import org.mrgeo.mapreduce.job.JobCancelledException;
import org.mrgeo.mapreduce.job.JobFailedException;
import org.mrgeo.opimage.MrsPyramidDescriptor;
import org.mrgeo.progress.Progress;

import java.io.IOException;
import java.util.Vector;

public class ChangeClassificationMapOp extends RasterMapOp
{
  private RasterMapOp source = null;

  private Classification classification;
  Aggregator aggregator = new MeanAggregator();

  @Override
  public void addInput(final MapOp n) throws IllegalArgumentException
  {
    if (source == null)
    {
      if (!(n instanceof RasterMapOp))
      {
        throw new IllegalArgumentException("Only raster inputs are supported for the source input.");
      }
      source = (RasterMapOp)n;
      _inputs.add(n);
    }
  }

  public static String[] register()
  {
    return new String[] { "changeClassification" };
  }

  @Override
  public void moveOutput(final String toName) throws IOException
  {
    super.moveOutput(toName);
    _outputName = toName;
    MrsImageDataProvider dp = DataProviderFactory.getMrsImageDataProvider(_outputName,
        AccessMode.READ, getProviderProperties());
    _output = MrsPyramidDescriptor.create(dp);
  }

  @Override
  public Vector<ParserNode> processChildren(final Vector<ParserNode> children, final ParserAdapter parser)
  {
    final Vector<ParserNode> result = new Vector<ParserNode>();

    if (children.size() < 2)
    {
      throw new IllegalArgumentException(
          "ChangeClassificationMapOp usage: changeClassification(source raster, type, [aggregation type])");
    }

    result.add(children.get(0));

    String cls = parseChildString(children.get(1), "classification type", parser);
    if (cls.equalsIgnoreCase("categorical"))
    {
      classification = Classification.Categorical;
    }
    else
    {
      classification = Classification.Continuous;
    }
    
    if (children.size() > 2)
    {
      if (children.size() == 3)
      {
        String agg = parseChildString(children.get(2), "aggregation type", parser);
        Class<? extends Aggregator> aggregatorClass = AggregatorRegistry.aggregatorRegistry.get(agg.toUpperCase());
        if (aggregatorClass != null)
        {
          try
          {
            aggregator = aggregatorClass.newInstance();
          }
          catch (InstantiationException e)
          {
            throw new IllegalArgumentException("Invalid aggregator " + agg);
          }
          catch (IllegalAccessException e)
          {
            throw new IllegalArgumentException("IllegalAccessException while instantiating aggregator " + agg);
          }
        }
      }
      else
      {
        throw new IllegalArgumentException(
            "ChangeClassificationMapOp usage: changeClassification(source raster, type, [aggregation type])");
      }
    }

    return result;
  }

  @Override
  public String toString()
  {
    return String.format("changetype " + (source == null ? "null" : source.toString()));
  }

  @Override
  public void build(final Progress p) throws IOException, JobFailedException, JobCancelledException
  {
    if (p != null)
    {
      p.starting();
    }

    final MrsImagePyramid sourcepyramid = RasterMapOp.flushRasterMapOpOutput(source, 0);
    
    MrsImageDataProvider provider = DataProviderFactory.getMrsImageDataProvider(sourcepyramid.getName(),
        AccessMode.READ, getProviderProperties());
    MrsImagePyramidMetadata metadata = provider.getMetadataReader().read();

    _outputName = source.getOutputName();
    
    if (metadata.getClassification() != classification)
    {
      metadata.setClassification(classification);
      metadata.setResamplingMethod(AggregatorRegistry.aggregatorRegistry.inverse().get(aggregator.getClass()));
      
      provider.getMetadataWriter().write();

      if (sourcepyramid.hasPyramids())
      {
        try
        {
          BuildPyramidSpark.build(sourcepyramid.getName(),
              aggregator, createConfiguration(), p, jobListener, getProviderProperties());
        }
        catch (Exception e)
        {
          e.printStackTrace();
        }
      }
    }

    MrsImageDataProvider dp = DataProviderFactory.getMrsImageDataProvider(_outputName,
        AccessMode.READ, getProviderProperties());
    _output = MrsPyramidDescriptor.create(dp);

    if (p != null)
    {
      p.complete();
    }

  }

}

File: mrgeo-mapalgebra/mrgeo-mapalgebra-image/src/main/scala/org/mrgeo/mapalgebra/ChangeClassificationMapOp.scala
/*
 * Copyright 2009-2017. DigitalGlobe, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package org.mrgeo.mapalgebra

import java.io.{Externalizable, IOException, ObjectInput, ObjectOutput}

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import org.apache.spark.{SparkConf, SparkContext}
import org.mrgeo.aggregators.AggregatorRegistry
import org.mrgeo.data.rdd.RasterRDD
import org.mrgeo.image.MrsPyramidMetadata.Classification
import org.mrgeo.job.JobArguments
import org.mrgeo.mapalgebra.parser.{ParserException, ParserNode}
import org.mrgeo.mapalgebra.raster.RasterMapOp

object ChangeClassificationMapOp extends MapOpRegistrar {

  override def register:Array[String] = {
    Array[String]("changeClassification")
  }

  def create(raster:RasterMapOp, classification:String):MapOp =
    new ChangeClassificationMapOp(Some(raster), Some(classification))

  def create(raster:RasterMapOp, classification:String, aggregator:String):MapOp =
    new ChangeClassificationMapOp(Some(raster), Some(classification), Some(aggregator))

  override def apply(node:ParserNode, variables:String => Option[ParserNode]):MapOp =
    new ChangeClassificationMapOp(node, variables)
}

@SuppressFBWarnings(value = Array("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE"), justification = "Scala generated code")
class ChangeClassificationMapOp extends RasterMapOp with Externalizable {

  private var rasterRDD:Option[RasterRDD] = None

  private var inputMapOp:Option[RasterMapOp] = None
  private var classification:Option[Classification] = None
  private var aggregator:Option[String] = None

  override def rdd():Option[RasterRDD] = rasterRDD

  override def setup(job:JobArguments, conf:SparkConf):Boolean = true

  override def execute(context:SparkContext):Boolean = {
    val input:RasterMapOp = inputMapOp getOrElse (throw new IOException("Input MapOp not valid!"))

    val meta = input.metadata() getOrElse
               (throw new IOException("Can't load metadata! Ouch! " + input.getClass.getName))

    rasterRDD = Some(
      input.rdd() getOrElse (throw new IOException("Can't load RDD! Ouch! " + inputMapOp.getClass.getName)))

    meta.setClassification(
      classification.getOrElse(throw new IOException("Can't get classification! Ouch! " + input.getClass.getName)))

    if (aggregator.isDefined) {
      meta.setResamplingMethod(aggregator.get)
    }

    metadata(meta)
    true
  }

  override def teardown(job:JobArguments, conf:SparkConf):Boolean = true

  override def readExternal(in:ObjectInput):Unit = {}

  override def writeExternal(out:ObjectOutput):Unit = {}

  private[mapalgebra] def this(raster:Option[RasterMapOp], classification:Option[String],
                               aggregator:Option[String] = None) = {
    this()
    inputMapOp = raster

    this.aggregator = aggregator match {
      case Some(s) =>
        val clazz = AggregatorRegistry.aggregatorRegistry.get(s.toUpperCase)
        if (clazz != null) {
          Some(s.toUpperCase)
        }
        else {
          throw new ParserException("Invalid aggregator " + s)
        }
      case _ => None
    }

    this.classification = Some(classification match {
      case Some(s) => s.toLowerCase match {
        case "categorical" => Classification.Categorical
        case _ => Classification.Continuous
      }
      case _ => throw new ParserException("Can't decode string")
    })

  }

  private[mapalgebra] def this(node:ParserNode, variables:String => Option[ParserNode]) = {
    this()

    if ((node.getNumChildren < 2) || (node.getNumChildren > 3)) {
      throw new ParserException(
        "ChangeClassificationMapOp usage: changeClassification(source raster, classification, [aggregation type])")
    }

    inputMapOp = RasterMapOp.decodeToRaster(node.getChild(0), variables)

    classification = Some(MapOp.decodeString(node.getChild(1), variables) match {
      case Some(s) => s.toLowerCase match {
        case "categorical" => Classification.Categorical
        case _ => Classification.Continuous
      }
      case _ => throw new ParserException("Can't decode string")
    })


    if (node.getNumChildren == 3) {
      aggregator = Some(MapOp.decodeString(node.getChild(2)) match {
        case Some(s) =>
          val clazz = AggregatorRegistry.aggregatorRegistry.get(s.toUpperCase)
          if (clazz != null) {
            s.toUpperCase
          }
          else {
            throw new ParserException("Invalid aggregator " + s)
          }
        case _ => throw new ParserException("Can't decode string")
      })
    }
  }

}
